chapter_generative-adversarial-networks/dcgan.md

```diff
@@ -407,16 +543,71 @@ def train(net_D, net_G, data_iter, num_epochs, lr, latent_dim,
           f'{metric[2] / timer.stop():.1f} examples/sec on {str(device)}')
 ```
 
+```{.python .input}
+#@tab tensorflow
+def train(net_D, net_G, data_iter, num_epochs, lr, latent_dim, device=d2l.try_gpu()):
+    loss = tf.keras.losses.BinaryCrossentropy(from_logits=True, reduction=tf.keras.losses.Reduction.SUM)
+    
+    for w in net_D.trainable_variables:
+        w.assign(tf.random.normal(mean=0, stddev=0.02, shape=w.shape))
+    for w in net_G.trainable_variables:
+        w.assign(tf.random.normal(mean=0, stddev=0.02, shape=w.shape))
+    
+    optimizer_hp = {"lr": lr, "beta_1": 0.5, "beta_2": 0.999}
+    optimizer_D = tf.keras.optimizers.Adam(**optimizer_hp)
+    optimizer_G = tf.keras.optimizers.Adam(**optimizer_hp)
+    
+    animator = d2l.Animator(xlabel='epoch', ylabel='loss', xlim=[1, num_epochs], nrows=2,
+                            figsize=(5, 5), legend=['discriminator', 'generator'])
+    animator.fig.subplots_adjust(hspace=0.3)
+    
+    for epoch in range(1, num_epochs + 1):
+        # Train one epoch
+        timer = d2l.Timer()
+        metric = d2l.Accumulator(3) # loss_D, loss_G, num_examples
+        for X, _ in data_iter:
+            batch_size = X.shape[0]
+            Z = tf.random.normal(mean = 0, stddev = 1, shape = (batch_size, 1, 1, latent_dim))
+            metric.add(d2l.update_D(X, Z, net_D, net_G, loss, optimizer_D),
+                       d2l.update_G(Z, net_D, net_G, loss, optimizer_G),
+                       batch_size)
+            
+        # Show generated examples
+        Z = tf.random.normal(mean=0, stddev=1, shape=(21, 1, 1, latent_dim))
+        # Normalize the synthetic data to N(0, 1)
+        fake_x = net_G(Z) / 2 + 0.5
+        imgs = tf.concat([tf.concat([fake_x[i * 7 + j] for j in range(7)], axis=1) 
+                          for i in range(len(fake_x) // 7)], axis=0)
+        animator.axes[1].cla()
+        animator.axes[1].imshow(imgs)
+        # Show the losses
+        loss_D, loss_G = metric[0] / metric[2], metric[1] / metric[2]
+        animator.add(epoch, (loss_D, loss_G))
+    print(f'loss_D {loss_D:.3f}, loss_G {loss_G:.3f}, '
+          f'{metric[2] / timer.stop():.1f} examples/sec on {str(device)}')
+```
+
 We train the model with a small number of epochs just for demonstration.
 For better performance,
 the variable `num_epochs` can be set to a larger number.
 
 ```{.python .input}
-#@tab all
 latent_dim, lr, num_epochs = 100, 0.005, 20
 train(net_D, net_G, data_iter, num_epochs, lr, latent_dim)
 ```
 
+```{.python .input}
+#@tab pytorch
+latent_dim, lr, num_epochs = 100, 0.005, 20
+train(net_D, net_G, data_iter, num_epochs, lr, latent_dim)
+```
+
+```{.python .input}
+#@tab tensorflow
+latent_dim, lr, num_epochs = 100, 0.0005, 40  # We have changed the learning rate and number of epochs 
```

Comment:
Why do we choose a different set of HPs?

Reply:
I tried several times using the shared learning rate and number of epochs. But I couldn't get any reasonable looking generated images in any of the trials. However, if we change the learning rate and number of epochs to the values that we have used (0.0005, and 40 respectively), almost always we get reasonable looking generated images. That's the reason why I changed those for TF.

Reply:
Can you get similar results to those produced by other frameworks when using the same HPs?